tests/unit_tests/sparse_blas/include/common_sparse_reference.hpp

```diff
@@ -0,0 +1,230 @@
+    std::vector<intType> jopa;
+    std::vector<fpType> opa;
+    if (transpose_val == oneapi::mkl::transpose::nontrans) {
+        iopa.assign(ia, ia + a_nrows + 1);
+        jopa.assign(ja, ja + nnz);
+        opa.assign(a, a + nnz);
+    }
+    else if (transpose_val == oneapi::mkl::transpose::trans ||
+             transpose_val == oneapi::mkl::transpose::conjtrans) {
+        iopa.resize(static_cast<std::size_t>(a_ncols + 1));
+        jopa.resize(nnz);
+        opa.resize(nnz);
+        do_csr_transpose(transpose_val, iopa.data(), jopa.data(), opa.data(), a_nrows, a_ncols,
+                         indexing, ia, ja, a);
+    }
+    else {
+        throw std::runtime_error("unsupported transpose_val=" +
+                                 std::to_string(static_cast<char>(transpose_val)));
+    }
+    return std::make_tuple(iopa, jopa, opa);
+}
+
+template <typename fpType>
+auto dense_transpose_if_needed(const fpType *x, std::size_t outer_size, std::size_t inner_size,
+                               std::size_t ld, oneapi::mkl::transpose transpose_val) {
+    std::vector<fpType> opx;
+    if (transpose_val == oneapi::mkl::transpose::nontrans) {
+        opx.assign(x, x + outer_size * ld);
+    }
+    else {
+        opx.resize(outer_size * ld);
+        for (std::size_t i = 0; i < outer_size; ++i) {
+            for (std::size_t j = 0; j < inner_size; ++j) {
+                opx[i + j * ld] = x[i * ld + j];
+            }
+        }
+    }
+    return opx;
+}
+
+/// Convert the sparse matrix in the given format to a dense matrix A in row major layout applied with A_view.
+template <typename fpType, typename intType>
+std::vector<fpType> sparse_to_dense(sparse_matrix_format_t format, const intType *ia,
+                                    const intType *ja, const fpType *a, std::size_t a_nrows,
+                                    std::size_t a_ncols, std::size_t nnz, intType indexing,
+                                    oneapi::mkl::transpose transpose_val,
+                                    oneapi::mkl::sparse::matrix_view A_view) {
+    oneapi::mkl::sparse::matrix_descr type_view = A_view.type_view;
+    oneapi::mkl::uplo uplo_val = A_view.uplo_view;
+    const bool is_symmetric_or_hermitian_view =
+        type_view == oneapi::mkl::sparse::matrix_descr::symmetric ||
+        type_view == oneapi::mkl::sparse::matrix_descr::hermitian;
+    // Matrices are not conjugated if they are symmetric
+    const bool apply_conjugate =
+        !is_symmetric_or_hermitian_view && transpose_val == oneapi::mkl::transpose::conjtrans;
+    std::vector<fpType> dense_a(a_nrows * a_ncols, fpType(0));
+
+    auto write_to_dense_if_needed = [&](std::size_t a_idx, std::size_t row, std::size_t col) {
+        if ((type_view == oneapi::mkl::sparse::matrix_descr::triangular ||
+             is_symmetric_or_hermitian_view) &&
+            ((uplo_val == oneapi::mkl::uplo::lower && col > row) ||
+             (uplo_val == oneapi::mkl::uplo::upper && col < row))) {
```

Comment:
Shouldn't this be the following?
```suggestion
            ((uplo_val == oneapi::mkl::uplo::lower && col >= row) ||
             (uplo_val == oneapi::mkl::uplo::upper && col <= row))) {
```

Reply:
Oh, nevermind. I see what you are doing. What you are doing is fine.